### ISSUE TITLE
Bug: Extra blank lines inserted in code blocks

### ISSUE THREAD

**Issue by 3kami3** [mattermost/mattermost-server#22891]
#### Summary
The bug is that extra blank lines are inserted when lists (1.,2.,3.,,) and code blocks (```) are used together.

#### VERSIONS
mattermost/mattermost-team-edition:7.9.1

#### OPARATION
Enter the following markdown code block as a message input.
````
1. test
```patch
1

3
```
````
Languages are reproduced not only in 'Diff(patch)' but also in 'Bash', etc.

#### Expected behavior
More unnecessary blank lines.

#### Observed behavior (that appears unintentional) 
No more unnecessary blank lines.
This bug affects not only the display but also the copy code.
Copying and pasting the patch code is an unexpected error factor.
<img width="458" alt="mm123" src="https://user-images.githubusercontent.com/12627100/230610237-57d9528f-5624-4f5b-b220-0524285af983.png">
<img width="458" alt="mm1_3" src="https://user-images.githubusercontent.com/12627100/230610259-048195a6-4506-4fe5-924d-fa4882444049.png">

**Comment by amyblais** [comment]
Opened https://mattermost.atlassian.net/browse/MM-52081.

**Comment by sugamadhiakri** [comment]
@amyblais could you assign this issue to me?

**Comment by sinansonmez** [comment]
@sugamadhiakri are you still working in this issue?

**Comment by adityasoni2019** [comment]
Hey can I take this up? @amyblais

**Comment by adityasoni2019** [comment]
Hey @amyblais, It seems to me like there's some issue with how the text is parsed by the "marked" library for markdown content.

As you can see in the below file - 

![image](https://github.com/user-attachments/assets/b617ef0c-9553-41c1-846b-7d6ccbfe8f39)

the output is - '1. test\n```patch\n1\n\n3\n```' (which is fine, cause it has 1 new line b/w 1 & 3 - 1 & 3)

but, as soon as the Makrdown.format() function is called (which is calling some functions that have input given by the markdown lib at some point in it's journey), the output changes to - 

![image](https://github.com/user-attachments/assets/682fa07d-76a6-45a9-b451-c2b9cb018381)

Which has 2 lines in b/w 1 & 3 

See below how the code function is called with - '1\n\n\n3' (2 lines b/w 1 & 3; and this input has been I think given by the markdown file)

![image](https://github.com/user-attachments/assets/496262e1-ffbe-4c88-84b9-0cb6f6d95b1e)


So, if there's any dev that has some perspective on this, that'd be great! 

Thanks!

**Comment by amyblais** [comment]
> It seems to me like there's some issue with how the text is parsed by the "marked" library for markdown content. . . . 

@3kami3 Feel free to come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on the Mattermost Community server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers Server" community channel](https://community.mattermost.com/core/channels/developers-server).

**Comment by hmhealey** [comment]
Taking a quick look through that, I noticed this commit where I made some changes to how whitespace in code blocks are handled:

https://github.com/mattermost/marked/commit/cd82cf9e4d1b1a8bcfc1389e9d2f13ac7800e541

That sounds like it could be related to this issue. It was a fix for https://github.com/mattermost/mattermost/issues/6865

**Comment by adityasoni2019** [comment]
Thanks for looking into it, @hmhealey. I was wondering if there's a way to put debuggers on the marked library codebase, while running my mattermost/mattermost? If not, any recos on how to debug it?

**Comment by hmhealey** [comment]
If you compile it locally, you should be able to add `debugger;` line to set a breakpoint from the code that will be hit by the browser's dev tools.

And yeah, you can change that to your marked fork. During development, you might want to change it to `file:` URL to have it use your local copy, and then once you're ready to submit PRs, you can change it to point to your fork while we test it. Once the marked changes are merged, then we'll make one last change to point it back to our fork.

**Comment by adityasoni2019** [comment]
Hey @hmhealey, I looked into the marked library, and the issue is certainly there.  I'll be using the same string being used for test in the description (1. test patch 1   3 )

Consider this piece of code from the marked library. (I've added some comments). 

```
if (cap = this.rules.list.exec(src)) { 

      // THIS CONDITION CHECKS IF THE SRC IS A LIST
      // THE VALUE OF CAP HERE COMES OUT TO BE AN ARRAY. AND cap[0] = '1. test\n```patch\n1\n\n' 
      // (Notice that there are already 2 new lines here)

     ...someCode

      for (; i < l; i++) {
        item = cap[i]; // ITEM = '1. test\n```patch\n1\n\n'

       ...someCode

        var newlinesInItem = /^([^\n]+)\n(\S[\s\S]+)$/.exec(item);
        // newlinesInItem COMES OUT TO BE AN ARRAY. 

        if (newlinesInItem && newlinesInItem.length > 0) {

          src = newlinesInItem[2] + '\n' + cap.slice(i + 1).join('\n') + src;
          // newlinesInItem[2] = '```patch\n1\n\n'
    

          // NOW, HERE ^ WHILE FORMING THE SRC, WE'RE ADDING ANOTHER '\n' AT THE END OF  newlinesInItem[2], 
          // WHICH ALREADY HAS 2 NEW LINES. WHICH IS THE ISSUE. 

          l = i + 1;
        }
```
So, basically, either we can fix the `this.rules.list.exec(src)` so, that in the value of cap, there's only 1 new line, cause this is where the 2 \n are originating from. OR, we can do something about how this line is handled - `src = newlinesInItem[2] + '\n' + cap.slice(i + 1).join('\n') + src;`

I'm reading more about this, but it would help to know your take on this; especially what was kept in mind while writing the logic for `this.rules.list` and adding the `\n` after `newlinesInItem[2]`

PS: The value of `this.rules.list = /^( *)((?:[*+-]|\d+\.)) [\s\S]+?(?:\n+(?=\1?(?:[-*_] *){3,}(?:\n+|$))|\n+(?= *\[([^\]]+)\]: *<?((?:\w*?:(?:\/\/)?|\/|#)[^\s>]+)>?(?: +["(]([^\n]+)[")])? *(?:\n+|$))|\n{2,}(?! )(?!\1(?:[*+-]|\d+\.) )\n*|\s*$)/` 

Thanks.

**Comment by hmhealey** [comment]
Oh, wow. Great find!

I didn't expect that the list is somehow the cause of this. I'm not super surprised that past-me the cause of this bug though 😅

This code was added in https://github.com/mattermost/marked/pull/4 (MM-1319) and https://github.com/mattermost/marked/pull/6 (MM-3192). For context, MM-1319 was to make it so that
```md
1. This is a list
This is some other text
```

would appear as 

> 1. This is a list
>
> This is some other text

instead of making "This is some other text" part of the list item like

> 1. This is a list
> This is some other text

which Markdown normally does.

I think I did it this way because changing `this.rules.text` was too difficult, and I still feel that way. It's incredibly easy to break those incredibly complicated regexes. Ideally, it wouldn't that wouldn't use a regular expression, but that's obviously way outside the scope of this issue.

I think we should look at the whole block involving `newlinesInItem`. I don't know if the problem on that `src = newlinesInItem[2]...` line is the initial `+ '\n'` or if it's caused by the `.join('\n')`, but that line is definitely the cause here

**Comment by adityasoni2019** [comment]
It's not caused by `.join('\n')` (and it breaks when I remove the join function) and everything's working fine when I remove the `+ '\n'`

May I know the reason behind adding the '\n' here?

<img width="1582" alt="image" src="https://github.com/user-attachments/assets/fe7e661a-4b2d-4156-ab9f-f1cf68c01472">

Thanks!

**Comment by hmhealey** [comment]
Thanks for the second ping since I missed the first one!

I don't actually remember why that is. If it doesn't cause MM-3192 to reappear, I'd be up for removing it.

Worst case scenario, if removing it ends up causing other issues that we don't catch, then we can always revisit this again. It'll be more evidence that we need to switch off our marked.js fork and onto the commonmark.js fork that we use in the mobile app, and I've wanted to make that change for years now 😅
